### PR TITLE
Fixing a NPE with failure policy=continue and base class has failure

### LINKED ISSUE #990: Invoker throws NullPointerExceptions after a configuration method fails in a test's super class when configfailurepolicy="continue"
Here is a simple test that extends a test base class:

``` java
public abstract class TestBase {

    @BeforeTest
    protected void setup() {
        System.out.println("Starting test");
        throw new RuntimeException("Fail the test.");
    }

    @Test
    protected abstract void execute();

    @AfterTest(alwaysRun = true)
    protected void cleanup() {
        System.out.println("This should always run");
    }
}

public class MyTest extends TestBase {

    @Override
    @Test
    protected void execute() {
        System.out.println("Test");
    }

}
```

I run MyTest with configfailurepolicy set to "continue". Setup() fails as expected, but cleanup() doesn't run despite alwaysRun=true.

This is whats happening:
1. Setup() fails, causing Invoker.setClassInvocationFailure() to be called. When setClassInvocationFailure() is called, the TestBase class is used as the key for m_classInvocationResults.
2. When Invoker.confInvocationPassed() is called for execute() and cleanup(), m_classInvocationResults.get() is called each time (line 426). The get() is called using MyTest as the key, but TestBase was the only key added to the map.
3. Null is returned, causing the contains() call to throw a NullPointerException.
4. The exception is caught at line 672, and the method invocation fails.

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #990

Fixes #990  .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: src/main/java/org/testng/internal/Invoker.java [hunk 1]
```diff
@@ -430,8 +430,24 @@ private boolean confInvocationPassed(ITestNGMethod method, ITestNGMethod current
...
         } else {
-          synchronized(m_classInvocationResults){
-             result = !m_classInvocationResults.get(cls).contains(instance);
+          synchronized (m_classInvocationResults) {
+            Set<Object> set = m_classInvocationResults.get(cls);
+            //If the set was null, it means that the test method is residing in a child class
+            //and maybe the parent method has configuration methods which may have had a failure
+            //So lets walk up the inheritance tree until either we find failures or till we
+            //reached the Object class.
+            while (set == null || cls.equals(Object.class)) {
```
Comment (juherr) [inline src/main/java/org/testng/internal/Invoker.java]: `while (set == null || !cls.equals(Object.class))`, true?

**Review by juherr — CHANGES_REQUESTED** [review]

File: src/main/java/org/testng/internal/Invoker.java (hunk 1 above)
Comment (krmahadevan) [reply src/main/java/org/testng/internal/Invoker.java]: My bad!! Fixed, amended commit and re-pushed. Please check

**Review by juherr — APPROVED** [review]

**Comment by krmahadevan** [comment]
ping @cbeust - Please let me know if there is anything required from side to get this merged.

**Comment by krmahadevan** [comment]
Resolved merge conflicts.

**Pull request merged**